internal/storage/streamed_chunk_uploader.go

```diff
@@ -0,0 +1,206 @@
+package storage
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+
+	"cloud.google.com/go/storage"
+	"github.com/googlecloudplatform/gcsfuse/internal/logger"
+	"github.com/googlecloudplatform/gcsfuse/internal/storage/gcs"
+	"github.com/googlecloudplatform/gcsfuse/internal/storage/storageutil"
+	"google.golang.org/api/googleapi"
+)
+
+type UploaderState int
+
+const (
+	// Default initial state in case of failure in initialization.
+	Uninitialized UploaderState = iota
+	// State of successful initialization, no writes so far.
+	Initialized
+	// No write is in progress, but there have been writes in the past.
+	Waiting
+	// Asynchronous write is in progress.
+	Writing
+	// Upload failed.
+	WriteError
+	// Has been finalized.
+	Destroyed
+)
+
+// A StreamedChunkUploader is an implementation of ChunkUploader
+// interface, which use storage.Writer from go-storage-client
+// for resumable upload.
+//
+// It also stores state of how many bytes have been sent to it
+// to write, how many have been successfully written by it so far,
+// and the current state of the uploader.
+type StreamedChunkUploader struct {
```

Comment:
Why should this struct be public. As the responsibility of this struct it to implement the chunkUploader interface, the init method should just return the interface type.

Reply:
Made private.